src/plugins/kibana_utils/public/state_management/url/url_tracker.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { createBrowserHistory, Location, History, createLocation } from 'history';
+import { useLayoutEffect } from 'react';
+
+/**
+ * Replicates what src/legacy/ui/public/chrome/api/nav.ts did
+ * Persists the url in sessionStorage so it could be restored if navigated back to the app
+ */
+export function createUrlTracker(key: string, history: History = createBrowserHistory()) {
+  return {
+    restoreUrl() {
+      const preservedLocation = JSON.parse(sessionStorage.getItem(key)!) as Location<any>;
+      if (preservedLocation) {
+        history.replace(preservedLocation);
+      }
+    },
+    startTrackingUrl() {
+      return history.listen(location => {
+        sessionStorage.setItem(key, JSON.stringify(location));
+      });
+    },
+    trackUrl(url: string | Location<any>) {
+      if (typeof url === 'string') {
+        url = createLocation(url);
+      }
+      sessionStorage.setItem(key, JSON.stringify(location));
+    },
+  };
+}
+
+export function useUrlTracker(appInstanceId: string, history: History) {
```

Comment:
That goes into the right direction - an addition here is that this method should only attempt to restore the URL if the current URL is `/` (or some other configurable "initial" URL). If the user copy/pastes a URL they got from Slack or something into the address bar, it should not get reset to their last active URL.

Also it should somehow replicate the `injectNewGlobalState` behavior - if you switch from visualize to dashboard, change the time range there and go back to visualize, the time range should not get "restored" to the value it had when visualize was last active - the value from the `data` plugin should take precedence (at least that's the current behavior).

Reply:
Thanks for pointing those 2 out. will try to figure something out

Reply:
@flash1293 

1. I just added an optional "shouldRestoreUrl" callback. So app can decide if the url can be restored. In my example app, as you suggested, I am checking for "initial root". 
https://github.com/elastic/kibana/pull/53582/files#diff-a59c3f0b20c3a7d99696d597e58f7e09R51
https://github.com/elastic/kibana/pull/53582/files#diff-cecab2f52c21a882929689f170ad3ef8R174

2. This works due to `stateSync` util itself + sessionStorage sync strategy. 
For demo I created `globalStateContainer` which is just text - https://github.com/elastic/kibana/pull/53582/files#diff-cecab2f52c21a882929689f170ad3ef8R56
Then I made sure I am synching it's state to sessionStorage - https://github.com/elastic/kibana/pull/53582/files#diff-cecab2f52c21a882929689f170ad3ef8R207
So assuming we have app1 and app2 which write to the same key in sessionStorage. After leaving app1, app2 will pick up `_g` state from session storage on start up.

Here is a demo where I tried to capture those 2 behaviours: https://drive.google.com/open?id=1q4RH75u59FhMn83d4a2Q5woI4TbiMVu3

1. There are 2 instances of todo app 
2. todos state is scoped to each app (`_todo-1` `_todo-2`). It is synced to url and also `urlTracker` restores it if we navigate back to app by initialPath
3. `/` `/completed` /not-completed` are just react-router routes. They are restored by that `url-tracker` util also.
4. Text input in the bottom is `_g`. That one is shared between apps. Restored by `stateSync` util.

Reply:
Hey @Dosant, great work! One question: Does this mean the session storage value for `_g` will always take precedence over url-`_g`? If that's the case what would happen when a user has Kibana already opened in the browser with another app and copy/pastes another url with a `_g` parameter in the URL and hits enter? The session storage would be there and overwrite the pasted URL `_g` state, right? AFAIK currently the URL state would win in this scenario.

Reply:
Discussed offline:
Solved this by creating new syncStrategy: `KbnGlobalStateSyncStrategy` 

This is a wrapper around KbnUrlSyncStrategy and SessionStorageSyncStrategy which takes care of where to get initial state from.

Reply:
Some important updates: 
https://github.com/elastic/kibana/pull/53582#issuecomment-571086653
`KbnGlobalStateSyncStrategy` is no longer relevant.